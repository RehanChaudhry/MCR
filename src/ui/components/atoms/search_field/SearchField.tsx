import { COLORS, FONTS, FONT_SIZE, SPACE } from "config";
import { moderateScale } from "config/Dimens";
import useEffectWithSkipFirstTime from "hooks/useEffectWithSkipFirstTime";
import React, { useState } from "react";
import { usePreferredTheme } from "hooks";
import Search from "assets/images/search_icon.svg";
import Cross from "assets/images/cross.svg";
import {
  StyleProp,
  StyleSheet,
  TextInput,
  TextStyle,
  TouchableOpacity,
  View,
  ViewStyle
} from "react-native";
import { Color } from "react-native-svg";
import { optimizedMemo } from "ui/components/templates/optimized_memo/optimized_memo";

interface OwnProps {
  style?: StyleProp<ViewStyle>;
  textStyle?: StyleProp<TextStyle>;
  placeholder: string;
  onChangeText: (textToSearch?: string) => void;
  searchIcon: boolean;
  iconColor?: Color;
  clearIcon: boolean;
  shouldNotOptimize?: boolean;
}

type Props = OwnProps;

const SearchField = optimizedMemo<Props>(
  ({
    placeholder,
    style,
    textStyle,
    onChangeText,
    searchIcon,
    iconColor,
    clearIcon
  }) => {
    const [currentSearchText, setCurrentSearchText] = useState("");
    const theme = usePreferredTheme();

    useEffectWithSkipFirstTime(() => {
      const timeoutRef = setTimeout(() => {
        onChangeText(currentSearchText);
      }, 1500);

      return () => {
        clearTimeout(timeoutRef);
      };
    }, [currentSearchText, onChangeText]);

    return (
      <View
        style={[
          styles.container,
          {
            backgroundColor: theme.themedColors.primary
          },
          style
        ]}>
        {searchIcon && (
          <Search
            width={14}
            height={14}
            style={styles.leftIcon}
            testID={"left-icon"}
            fill={iconColor ?? theme.themedColors.interface[600]}
          />
        )}
        <TextInput
          value={currentSearchText}
          placeholderTextColor={theme.themedColors.interface[600]}
          placeholder={placeholder}
          numberOfLines={1}
          testID="SEARCH"
          style={[
            styles.textInput,
            { color: theme.themedColors.label },
            textStyle
          ]}
          onChangeText={setCurrentSearchText}
        />

        {clearIcon && currentSearchText !== "" && (
          <TouchableOpacity onPress={() => setCurrentSearchText("")}>
            <Cross
              width={14}
              height={14}
              testID={"right-icon"}
              style={styles.rightIcon}
              fill={iconColor ?? theme.themedColors.interface[600]}
            />
          </TouchableOpacity>
        )}
      </View>
    );
  }
);

const styles = StyleSheet.create({
  container: {
    height: 40,
    alignItems: "center",
    flexDirection: "row",
    justifyContent: "center",
    padding: SPACE.md,
    paddingTop: 2.5,
    paddingBottom: 2.5,
    borderRadius: 10
  },
  leftIcon: {
    width: moderateScale(12),
    aspectRatio: 1,
    marginRight: 8
  },
  rightIcon: {
    aspectRatio: 1,
    marginLeft: 8
  },
  textInput: {
    fontFamily: FONTS.regular,
    flex: 1,
    fontSize: FONT_SIZE.xsm,
    padding: 0,
    color: COLORS.textColor1
  },
  textInputSmallFont: {
    fontFamily: FONTS.regular,
    flex: 1,
    fontSize: FONT_SIZE._2xsm,
    padding: 0,
    color: COLORS.textColor1
  }
});

export default SearchField;
