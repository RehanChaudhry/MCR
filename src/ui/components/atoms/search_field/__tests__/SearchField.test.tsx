import { fireEvent, render, waitFor } from "@testing-library/react-native";
import React from "react";
import { SearchField } from "../SearchField";

it("Snapshot testing", () => {
  const rendered = render(
    <SearchField
      placeholder="Enter text"
      onChangeText={() => {}}
      leftIcon={false}
      rightIcon={false}
    />
  ).toJSON();
  expect(rendered).toMatchSnapshot();
});

it("Search by entered characters", async () => {
  const onChangeText = jest.fn();
  const { getByTestId } = render(
    <SearchField
      placeholder="Enter text"
      onChangeText={onChangeText}
      leftIcon={false}
      rightIcon={false}
    />
  );
  const textToSearch = getByTestId("SEARCH");

  await waitFor(() => {
    fireEvent.changeText(textToSearch, "Muhammad noman");
    expect(onChangeText).toBeCalledTimes(1);
    expect(onChangeText).toHaveBeenCalledWith("Muhammad noman");
  });
});

it("renders left icon", () => {
  const { queryByTestId } = render(
    <SearchField
      placeholder={"Enter Text"}
      onChangeText={() => {}}
      leftIcon={true}
      rightIcon={false}
    />
  );
  const leftIcon = queryByTestId("left-icon");
  const rightIcon = queryByTestId("right-icon");
  expect(rightIcon).toBeNull();
  expect(leftIcon).not.toBeNull();
});

it("renders right icon", () => {
  const { queryByTestId } = render(
    <SearchField
      placeholder={"Enter Text"}
      onChangeText={() => {}}
      leftIcon={false}
      rightIcon={true}
    />
  );
  const rightIcon = queryByTestId("right-icon");
  const leftIcon = queryByTestId("left-icon");
  expect(leftIcon).toBeNull();
  expect(rightIcon).not.toBeNull();
});
