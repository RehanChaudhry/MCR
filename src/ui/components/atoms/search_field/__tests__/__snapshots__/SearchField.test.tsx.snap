// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot testing 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "borderRadius": 10,
        "flexDirection": "row",
        "justifyContent": "center",
        "paddingBottom": 11,
        "paddingLeft": 12,
        "paddingRight": 12,
        "paddingTop": 11,
      },
      Object {
        "backgroundColor": "#00694e",
      },
      undefined,
    ]
  }
>
  <TextInput
    allowFontScaling={true}
    numberOfLines={1}
    onChangeText={[Function]}
    placeholder="Enter text"
    placeholderTextColor="#4B5563"
    rejectResponderTermination={true}
    style={
      Array [
        Object {
          "color": "#666666",
          "flex": 1,
          "fontFamily": "Inter-Regular",
          "fontSize": 25,
          "padding": 0,
        },
        Object {
          "color": "#111827",
        },
        undefined,
      ]
    }
    testID="SEARCH"
    underlineColorAndroid="transparent"
    value=""
  />
</View>
`;
